### PR TITLE
xrdp: 0.9.9 -> 0.9.21, xorgxrdp: 0.2.9 -> 0.9.19

### PR DESCRIPTION
**Pull request by chvp**
###### Description of changes

Package bump to get rid of known vulnerabilities. Includes rewrite and vendoring of the patch needed to make certain values configurable.

###### Things done

Tested by including this in my nixos config and using xrdp the way I usually do in my setup.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ajs124** [comment]
Probably supersedes #83922. Also related #83927.

**Comment by chvp** [comment]
Right, forgot to mention that. I've also seen the module refactor, but wanted to keep this PR as tight as possible to reduce friction in reviewing/merging. I was going to look into the module changes after this is merged.

**Comment by Mic92** [comment]
What desktop manager did you test this with? After applying your patch I only get a black screen after connecting:

```
$ wlfreerdp +clipboard /smart-sizing /multimon -grab-keyboard /ipv6 /v:$some-host:3389 /u:$someuser /p:$some-password /cert-ignore  "$@"
```

This is my xrdp configuration: https://github.com/TUM-DSE/doctor-cluster-config/blob/master/modules/xrdp.nix

The same configuration works with the old version.

**Comment by Mic92** [comment]
Also now it seems I have now to put `startwm.sh` into `/etc/xrdp/startwm.sh` for the desktop-manager to start.
@chvp is this an issue you want to address later?

```nix
  environment.etc."xrdp/startwm.sh" = {
    text = ''
      ${pkgs.runtimeShell}
      . /etc/profile
      ${pkgs.xfce.xfce4-session}/bin/xfce4-session
     '';
    mode = "755";
  };
```

**Comment by chvp** [comment]
Ah, I rebooted after deploying (I've noticed before that xrdp can be iffy after restarts due to upgrades but never looked into figuring out why). 

It's weird that the `startwm.sh` should be required, I've not noticed that and the NixOS test succeeds as well. I'll look into it sometime this week, because that seems like a regression.

**Comment by Mic92** [comment]
friendly ping.

**Comment by chvp** [comment]
Yes, sorry. I was going to look into it last week, but then my week turned out to be a lot busier than I expected. My hope is to look into it tomorrow evening.

**Comment by chvp** [comment]
There was a small part missing from the patch related to the startwm.sh functionality. This should be fixed now.

File: pkgs/applications/networking/remote/xrdp/default.nix [hunk 1]
```diff
@@ -58,23 +60,23 @@ let
       ./bootstrap
     '';
     dontDisableStatic = true;
-    configureFlags = [ "--with-systemdsystemunitdir=/var/empty" "--enable-ipv6" "--enable-jpeg" "--enable-fuse" "--enable-rfxcodec" "--enable-opus" ];
+    configureFlags = [ "--with-systemdsystemunitdir=/var/empty" "--enable-ipv6" "--enable-jpeg" "--enable-fuse" "--enable-rfxcodec" "--enable-opus" "--enable-pam-config=unix" ];
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/remote/xrdp/default.nix:63]:
```suggestion
    configureFlags = [
      "--with-systemdsystemunitdir=/var/empty"
      "--enable-ipv6"
      "--enable-jpeg"
      "--enable-fuse"
      "--enable-rfxcodec"
      "--enable-opus"
      "--enable-pam-config=unix"
    ];
```

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by Mic92** [comment]
Probably should be backported because we also marked xrdp on nixos 22.11 as insecure.

**Comment by kvtb** [comment]
Here:
https://github.com/nixos/nixpkgs/blob/1c272b106b66baea9dd8ef7cc2c541dd38373657/pkgs/applications/networking/remote/xrdp/default.nix#L31

`{xrdp.src}` should point to the xrdp sources after patch applied.

**Pull request merged**